presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java

```diff
@@ -459,9 +503,189 @@ else if (!result.getBlocked().isDone()) {
                 // get all sections ready for execution
                 .filter(this::isReadyForExecution)
                 .limit(maxConcurrentMaterializations - runningPlanSections)
+                .map(this::tryCostBasedOptimize)
                 .collect(toImmutableList());
     }
 
+    /**
+     * A general purpose utility function to invoke runtime cost-based optimizer.
+     * (right now there is only one plan optimizer which determines if the probe and build side of a JoinNode should be swapped
+     * based on the statistics of the temporary table holding materialized exchange outputs from finished children sections)
+     */
+    private StreamingPlanSection tryCostBasedOptimize(StreamingPlanSection section)
+    {
+        // no need to do runtime optimization if no materialized exchange data is utilized by the section.
+        if (!isRuntimeOptimizerEnabled(session) || section.getChildren().isEmpty()) {
+            return section;
+        }
+
+        // Apply runtime optimization on each StreamingSubPlan and generate optimized new fragments
+        Map<PlanFragment, PlanFragment> oldToNewFragment = new HashMap<>();
+        stream(forTree(StreamingSubPlan::getChildren).depthFirstPreOrder(section.getPlan()))
+                .forEach(subPlan -> performRuntimeOptimizations(subPlan, oldToNewFragment));
```

Comment:
Hmm. Now `performRuntimeOptimizations` updates `oldToNewFragment`. Ideally we want the input/output of methods to be more clear (e.g. pass in the current plan, returns the new optimized plan)

What about the following:

- Make `performRuntimeOptimizations` to return `Optional<PlanFragment>` as follows:
```java
    private Optional<PlanFragment> getRuntimeOptimizedPlan(StreamingSubPlan subPlan)
    {
        PlanFragment fragment = subPlan.getFragment();
        PlanNode newRoot = fragment.getRoot();
        for (PlanOptimizer optimizer : runtimePlanOptimizers) {
            newRoot = optimizer.optimize(newRoot, session, variableAllocator.getTypes(), variableAllocator, idAllocator, warningCollector);
        }
        // The partitioningScheme should stay the same
        // even if the root's outputVariable layout is changed.
        if (newRoot != fragment.getRoot()) {
            return Optional.of(
                    new PlanFragment(
                            fragment.getId(),
                            newRoot,
                            fragment.getVariables(),
                            fragment.getPartitioning(),
                            scheduleOrder(newRoot),
                            fragment.getPartitioningScheme(),
                            fragment.getStageExecutionDescriptor(),
                            fragment.isOutputTableWriterFragment(),
                            fragment.getStatsAndCosts(),
                            fragment.getJsonRepresentation())));
        }
    }
```

- Rewrite this line to 
```java
        stream(forTree(StreamingSubPlan::getChildren).depthFirstPreOrder(section.getPlan()))
                .forEach(currentPlan -> {
                    Optional<PlanFragment> newPlan = getRuntimeOptimizedPlan(currentPlan);
                    if (newPlan.isPresent()) {
                        oldToNewFragment.put(currentPlan.getFragment(), newPlan.get());
                    }
                });
```

This will still contain a multi-line anonymous lambda, but it should be easier to follow than the original one :)